pkg/apis/seedmanagement/v1alpha1/defaults.go

```diff
@@ -0,0 +1,103 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement/helper"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+// SetDefaults_ManagedSeed sets default values for ManagedSeed objects.
+func SetDefaults_ManagedSeed(obj *ManagedSeed) {
+	switch {
+	case obj.Spec.SeedTemplate != nil:
+		// Set seed defaults
+		seed := gardencorev1beta1.Seed{Spec: obj.Spec.SeedTemplate.Spec}
+		gardencorev1beta1.SetDefaults_Seed(&seed)
+		obj.Spec.SeedTemplate.Spec = seed.Spec
+
+		// Set default seed secret references
+		setDefaultSecretRef(&obj.Spec.SeedTemplate.Spec, obj)
+		setDefaultBackupSecretRef(&obj.Spec.SeedTemplate.Spec, obj)
+
+	case obj.Spec.Gardenlet != nil && obj.Spec.Gardenlet.Config != nil:
+		// Initialize gardenlet config
+		var gardenletConfig *configv1alpha1.GardenletConfiguration
+		var err error
+		if obj.Spec.Gardenlet.Config != nil {
+			gardenletConfig, err = helper.DecodeGardenletConfigExternal(obj.Spec.Gardenlet.Config)
+			if err != nil {
+				return
+			}
+		} else {
+			gardenletConfig = &configv1alpha1.GardenletConfiguration{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: configv1alpha1.SchemeGroupVersion.String(),
+					Kind:       "GardenletConfiguration",
+				},
+			}
+		}
+
+		// Set gardenlet config defaults
+		configv1alpha1.SetDefaults_GardenletConfiguration(gardenletConfig)
+
+		// Initialize seed config
+		if gardenletConfig.SeedConfig == nil {
+			gardenletConfig.SeedConfig = &configv1alpha1.SeedConfig{}
+		}
+
+		// If merging with parent is disabled, set seed defaults
+		gardencorev1beta1.SetDefaults_Seed(&gardenletConfig.SeedConfig.Seed)
+
+		// Set default seed secret references
+		setDefaultBackupSecretRef(&gardenletConfig.SeedConfig.Seed.Spec, obj)
+
+		// Marshal gardenlet config to JSON
+		gardenletConfigJSON, _ := json.Marshal(gardenletConfig)
+
+		// Set gardenlet config back to obj.Spec.Gardenlet.Config
+		obj.Spec.Gardenlet.Config = &runtime.RawExtension{Raw: gardenletConfigJSON}
+	}
+}
+
+func setDefaultSecretRef(seedSpec *gardencorev1beta1.SeedSpec, obj *ManagedSeed) {
+	if seedSpec.SecretRef == nil {
+		seedSpec.SecretRef = &corev1.SecretReference{
+			Name:      fmt.Sprintf("seed-%s", obj.Name),
+			Namespace: obj.Namespace,
+		}
+	}
+}
+
+func setDefaultBackupSecretRef(seedSpec *gardencorev1beta1.SeedSpec, obj *ManagedSeed) {
+	if seedSpec.Backup != nil && seedSpec.Backup.SecretRef == (corev1.SecretReference{}) {
```

Comment:
It's defaulted since from my perspective it makes a lot of sense to specify the default names here, rather than in the managed seed controller, since if `backup` is not nil, then this `secretRef` is mandatory. 

Note that the overall behavior is not changed compared to `use-as-seed`. The managed seed controller code has been changed to only create the backup secret if it doesn't exists and only update it if it's owned by the managed seed resource. This means that if the user specified their own secret, the controller won't touch it.

Reply:
> It's defaulted here now since from my perspective it makes much more sense to specify the default names here, rather than in the managed seed controller.

Yes, of course. This is definitely the way to go.

> The `use-as-seed` annotation implementation does default the backup secret ref if it's not specified.

Ok, I see. I've mixed up things here, you're right.